lib/src/util/side_effect_visitor.dart

```diff
@@ -0,0 +1,546 @@
+import 'package:analyzer/analyzer.dart';
+
+class SideEffectVisitor extends RecursiveAstVisitor {
+  var _hasSideEffect = false;
+
+  bool get hasSideEffect => _hasSideEffect;
+  @override
+  visitAdjacentStrings(AdjacentStrings node) {
+    super.visitAdjacentStrings(node);
+  }
+
+  @override
+  visitAnnotation(Annotation node) {
+    super.visitAnnotation(node);
+  }
+
+  @override
+  visitArgumentList(ArgumentList node) {
+    super.visitArgumentList(node);
+  }
+
+  @override
+  visitAsExpression(AsExpression node) {
+    super.visitAsExpression(node);
+  }
+
+  @override
+  visitAssertStatement(AssertStatement node) {
+    super.visitAssertStatement(node);
+  }
+
+  @override
+  visitAssignmentExpression(AssignmentExpression node) {
+    _hasSideEffect = true;
+  }
+
+  @override
+  visitAwaitExpression(AwaitExpression node) {
+    super.visitAwaitExpression(node);
+  }
+
+  @override
+  visitBinaryExpression(BinaryExpression node) {
+    super.visitBinaryExpression(node);
```

Comment:
Section 10.1.1 of the specification (http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-408.pdf) has a list.